### DISCUSSION TITLE
serious error in rfsrc.impute?

### DISCUSSION THREAD

**Discussion by drjhw** [kogalur/randomForestSRC#328] [Q&A]
I wanted to use rfsrc.impute to impute missing data in a large data-set (just over 1 million cases).

Because my data-set is so large, I wanted to test the basics of rfsrc.impute. So, I ran the multivariate example provided in the rfsrc.impute help "(suggested by John Sheffield)" - but I set **test_rows <- 1e5** (instead of **test_rows=1000**). rfsrc_impute runs OK. It reports:

multivariate missForest parameters: (#iter, #vars, #blks)=(10,9,5)
       iteration:1
.....
         >>  err = 0.952 ,  drop = Inf
       iteration:2
.....
         >>  err = 0.221 ,  drop = 0.731
       iteration:3
.....
         >>  err = 0.181 ,  drop = 0.0399
       iteration:4
.....
         >>  err = 0.187 ,  drop = -0.00575

The plotting routine then issues a dozen  warnings of the form:
1: In supsmu(x, y, span = 0.25) :
  3 observations with NAs, NaNs and/or Infs deleted
2: In supsmu(x, y, span = 0.25) :
  3 observations with NAs, NaNs and/or Infs deleted

The plots of the results looks awful (run it yourself and see - the image is too large to attach to this e-mail).

So, I computed the Spearman rank correlation coefficient for the values of fake_data$i and Imputed_data$i, where fake_data_missing$i==NA

imiss=is.na(fake_data_missing$i)
cor.test(fake_data$i[imiss],imputed_data$i[imiss],method='s')

      Spearman's rank correlation rho

data:  fake_data$i[imiss] and imputed_data$i[imiss]
S = 1.0624e+13, p-value = 0.1766
alternative hypothesis: true rho is not equal to 0
sample estimates:
        rho
0.006754003 

Next, I checked the initial 100 values of fake_data$i and imputed_data$i. They appear to correlate!!

cbind(fake_data$i[imiss][1:100],imputed_data$i[imiss][1:100])

           [,1]      [,2]
  [1,]  791.5054  768.3957
  [2,]  872.6438  824.4620
  [3,]  907.1745  850.7277
  [4,] 1010.1592 1032.1285
  [5,]  864.9064  876.7496
  [6,] 1157.1776 1087.8610
  [7,]  808.6953  777.4076
  [8,]  894.1729  740.5245
  [9,]  872.3655  852.8704
 [10,]  977.6915  982.8363
 [11,]  882.9193  964.6375
 [12,]  936.0603 1006.7200
 [13,]  981.6240 1058.0308
 [14,]  990.3182 1048.3447
 [15,]  920.3623  849.5419
 [16,]  899.4895  868.1647
 [17,] 1069.9544 1069.3912
 [18,]  952.3077 1051.5682
 [19,] 1094.8970  981.7823
 [20,]  714.0084  783.7396
etc.

Now, it is obvious from looking at these data that there is​ a correlation. I checked that:-

      Spearman's rank correlation rho
data:  fake_data$i[imiss][1:100] and imputed_data$i[imiss][1:100]
S = 23700, p-value < 2.2e-16
alternative hypothesis: true rho is not equal to 0
sample estimates:
      rho
0.8577858

The same is true for the next 300 pairs of points:-

      Spearman's rank correlation rho
data:  fake_data$i[imiss][101:400] and imputed_data$i[imiss][101:400]
S = 552770, p-value < 2.2e-16
alternative hypothesis: true rho is not equal to 0
sample estimates:
      rho
0.8771609

But, then the correlation breaks down, somewhere between the 400th and 500th pair of data:-

      Spearman's rank correlation rho
data:  fake_data$i[imiss][401:500] and imputed_data$i[imiss][401:500]
S = 55806, p-value < 2.2e-16
alternative hypothesis: true rho is not equal to 0
sample estimates:
      rho
0.6651305 

So, by the 500th - 600th pairs, there is no relationship

   Spearman's rank correlation rho
data:  fake_data$i[imiss][501:600] and imputed_data$i[imiss][501:600]
S = 160854, p-value = 0.7308
alternative hypothesis: true rho is not equal to 0
sample estimates:
       rho
0.03477948

Any random set of 100 pairs thereafter shows no relationship
e.g. 1501:1600
      Spearman's rank correlation rho
data:  fake_data$i[imiss][1501:1600] and imputed_data$i[imiss][1501:1600]
S = 174744, p-value = 0.6308
alternative hypothesis: true rho is not equal to 0
sample estimates:
        rho
-0.04856886

or 20001:20100
     Spearman's rank correlation rho
data:  fake_data$i[imiss][20001:20100] and imputed_data$i[imiss][20001:20100]
S = 171968, p-value = 0.7522
alternative hypothesis: true rho is not equal to 0
sample estimates:
        rho
-0.03191119

and the good correlation in the first 400 pairs is swamped by the lack of any correlation for the next 99600 pairs:-

      Spearman's rank correlation rho
data:  fake_data$i[imiss] and imputed_data$i[imiss]
S = 1.0624e+13, p-value = 0.1766
alternative hypothesis: true rho is not equal to 0
sample estimates:
        rho
0.006754003

So, something is going wrong between the 400th and 500th pair.

I then re-ran the example setting test_rows=2000, with no problem. But, when I set test_rows=1e4, I again receive the warnings from the plotting routine:-

Warning messages:
1: In supsmu(x, y, span = 0.25) :
  3 observations with NAs, NaNs and/or Infs deleted
2: In supsmu(x, y, span = 0.25) :
  3 observations with NAs, NaNs and/or Infs deleted
3: In supsmu(x, y, span = 0.25) :
  1 observation with NA, NaN or Inf deleted
etc.

With test_rows=1e4, the imputation appears fairly-to-very inaccurate:-

      Spearman's rank correlation rho
data:  fake_data$i[imiss] and imputed_data$i[imiss]
S = 8082809018, p-value < 2.2e-16
alternative hypothesis: true rho is not equal to 0
sample estimates:
      rho
0.2551591

As before, the imputation is initially accurate - but now for the first  800 pairs of data:-

      Spearman's rank correlation rho
data:  fake_data$i[imiss][801:900] and imputed_data$i[imiss][801:900]
S = 8674, p-value < 2.2e-16
alternative hypothesis: true rho is not equal to 0
sample estimates:
      rho
0.9479508

However, by the 1000th-1100th pairs, there is no longer any correlation:-

    Spearman's rank correlation rho
data:  fake_data$i[imiss][1001:1100] and imputed_data$i[imiss][1001:1100]
S = 156350, p-value = 0.5407
alternative hypothesis: true rho is not equal to 0
sample estimates:
       rho
0.06180618

So, again, something is going wrong in the process of imputation after it begins well.

Obviously, this major error may have serious consequences for people who - like me - want to impute missing data in large data-sets.

Could you please (a) indicate the nature of the problem and (b) provide a 'fix' so that I can take my analysis forward?

**Comment by ishwaran** [comment]
Hi, sorry for the slow reply but I finally tracked down the issue.  

What we have to remember is that the `impute` function first removes all rows which have complete missingness - i.e. where all columns are `NA`.  Here I copy point 5 from the help file:

 _5. Prior to imputation, the data is processed and records with all values missing are removed, as are variables having all missing values._

Therefore, what was happening is that the imputed data `imputed_data` had a different number of rows than the fake data `fake_data` which was leading to the strange results (also this is why the `supsmu` function was throwing out errors).

The fix is easy.  After making the data missing, add the following lines (I will make this correction to the help file to avoid other users having this same issue):

```
#remove rows where all the data is missing
all.r.na <- apply(fake_data_missing, 1, function(x){all(is.na(x))})
if (sum(all.r.na) > 0) {
  fake_data <- fake_data[!all.r.na,]
  fake_data_missing <- fake_data_missing[!all.r.na,]
}

```

**Comment by drjhw** [comment]
Great, thanks,

Jonathan

________________________________

**Comment by drjhw** [comment]
BTW - you seem to have missed my earlier e-mail, which I copy below:-

non-reproducible results with sidClustering
Jonathan Williams
Hello,

I need to be able to reproduce results reliably (doesn't everybody?).

I used the sidClustering() procedure to cluster about 200 microbiome measures and obtained a significant correlation between the resulting clusters and my patients' clinical diagnoses. But, when I repeated the same analysis later, the correlation was different!

In fact, the clusters (and hence the correlation) differ every time I run sidClustering() - even though I set R's random number seed twice - immediately before invoking sidClustering() and within the sidClustering() call itself:-

set.seed(111); s1=sidClustering(data, seed=-111L)
t1=table(s1$clus, diagnosis)
print(t1); chisq.test(t1) # t1 differs for every run; so does the resulting chi square

Please tell me how I can constrain sidClustering() to deliver reproducible results.

Thanks,

Jonathan Williams

PS - does rfsrc's 'seed' option really require a negative integer? rfsrc's 'manual' says that it does.

**Reply by ishwaran** [reply]
Yes, seed needs to be a negative number.

Reproducibility for the package cannot be guaranteed when variable importance (VIMP) is used.  The default action by `sidClustering` is to reduce the dimension space first using VIMP; in particular see the option `reduce` (help file information given below)

 _reduce: Apply dimension reduction? Uses holdout vimp which is computationally intensive and conservative but has good false
          discovery properties.  Only applies to ‘method="sid"’._

Therefore, unfortunately, if you want reproducibility via the seed option, you cannot use reduce.  The following should give reproducible results as an example:

```
o=sidClustering(mtcars,reduce=F,seed=-254)
o2=sidClustering(mtcars,reduce=F,seed=-254)
```

**Comment by drjhw** [comment]
Thanks. yes, that works. Sorry I didn't notice that the rfsrc help specifies that seed should be a negative integer.

But, now my next problem is to increase the number of trees for sidClustering(). The analysis only ever uses the default value of 500, irrespective of the number that I set using the "ntree=" option. Therefore, sid1$rf$ntree and sid2$rf$ntree are both 500:-
sid1=sidClustering(data=data, k=2, ntree=5000, seed=-1, reduce=F)
sid2=sidClustering(data=data, k=2, ntree=50, seed=-1, reduce=F)
sid1$rf$ntree
# [1] 500
sid2$rf$ntree
# [1] 500
setting ntreee= to a negative integer makes no difference.
This means that I can't grow a larger forest, to improve the precision of the sidClustering() analyses.

Thanks for your help
________________________________

**Comment by drjhw** [comment]
Strangely, if I set ntree.reduce explicitly, then ntree itself becomes effective:-

sid3=sidClustering(data=data, ntree=5e3, reduce=F, ntree.reduce=10, seed=-1)
sid3$rf$ntree # [1] 5000

I don't understand what is happening, here
________________________________

**Comment by ishwaran** [comment]
I think you are correct that this is a bug.  We will correct this in the next release of the package.  thank you for pointing this out.
